knr1050Sup/knr1050.db

```diff
@@ -7,182 +26,363 @@ record(bo, "$(P)SIM") {
+
+record(bo, "$(P)_PUMP:START:SP") {
+    field(DESC, "Start ramp")
+    field(DTYP, "stream")
     field(SCAN, "Passive")
-	
-	field(OUT, "@knr1050.proto ramp($(P),_FLOW:SP:CONV,CON:A:SP,CON:B:SP,CON:C:SP,CON:D:SP) $(PORT)")
-	field(ZNAM, "")
-	field(ONAM, "Start")
+    
+    field(OUT, "@knr1050.proto ramp($(P),_FLOW:SP:CONV,COMP:,A:SP,B:SP,C:SP,D:SP) $(PORT)")
+    field(ZNAM, "")
+    field(ONAM, "Start")
+    
+    field(FLNK, "$(P)_STOP:KLV:SP PP")
+    
+    field(DISV, "0")
+    field(SDIS, "$(P)DISABLE")
+}
 
-	field(FLNK, "$(P)_STOP:KLV:SP PP")
-	
-	field(DISV, "0")
-	field(SDIS, "$(P)DISABLE")
+record(bo, "$(P)_PUMP:STOP:SP") {
+    field(DESC, "Stops the pump")
+    field(DTYP, "stream")
+    field(SCAN, "Passive")
+    
+    field(OUT,  "@knr1050.proto stop_pump $(PORT)")
+    field(ZNAM, "")
+    field(ONAM, "Stop")
 }
 
-alias("$(P)PUMP:START:SP", "$(P)PUMP:START")
+alias("$(P)_PUMP:STOP:SP", "$(P)_PUMP:STOP")
 
 record(bo, "$(P)_STOP:KLV:SP") {
-	field(DESC, "Stop mode: keep last values")
-	field(DTYP, "stream")
+    field(DESC, "Stop mode: keep last values")
+    field(DTYP, "stream")
 
-	field(OUT,  "@knr1050.proto stop_klv $(PORT)")
+    field(OUT,  "@knr1050.proto stop_klv $(PORT)")
+    
+    field(FLNK, "$(P)TIME:START")
 }
 
 alias("$(P)_STOP:KLV:SP", "$(P)_STOP:KLV")
 
 ###########################################################
-# Get the device status
+# 4. TIMED RUN CONTROL
 ###########################################################
+#
+record(calc, "$(P)TIME")
+{
+    field(DESC, "Current run time in seconds")
+    field(EGU, "s")
+    field(SCAN, "1 second")
+    field(INPA, "$(P)TIME:STATUS")
+    
+    field(CALC, "A==0?0:VAL+1")
+    
+    field(FLNK, "TIME:CHECK")
+}
 
-# ///
-# /// The status is sent to a calculation record buffer. The 
-# /// records that relate to the status returns are then 
-# /// populated using the buffer elements. 
-# ///
-record(ai, "$(P)GET_STATUS") {
-	field(DESC, "Get the current device status")
-	field(DTYP, "stream")
-	field(SCAN, "1 second")
-	
-	field(INP, "@knr1050.proto get_status($(P)BUFFER) $(PORT)")
-	
-	info(alarm, "Knr1050")
+record(mbbi, "$(P)TIME:STATUS") 
+{
+    field(DESC, "Mode of time checking")
+    
+    field(ZRVL, "0")
+    field(ONVL, "1")
+    field(TWVL, "2")
+  
+    field(ZRST, "STOP")
+    field(ONST, "TIMED")
+    field(TWST, "CONSTANT")
+    
+    info(INTEREST, "HIGH")
 }
 
-record(calc, "$(P)BUFFER") {
-	field(DESC, "Parsed values of get_status")
-	field(CALC, "0")
+record(ai, "$(P)TIME:STAMP") {
+    field(DESC, "Device time stamp in ms")
+    field(EGU, "ms")
+    
+    field(INP, "$(P)BUFFER.A CP")
+}
+
+record(ao, "$(P)TIME:SP")
+{
+    field(DESC, "Desired pump runtime in s")
+    field(DTYP, "Soft Channel")
+    field(DRVL, "1")
+    field(DRVH, "1000000")
+    
+    field(EGU, "s")
+    field(FLNK, "$(P)TIME:CALCVOL")
+    info(INTEREST, "HIGH")
+}
+alias("$(P)TIME:SP", "$(P)TIME:SP:RBV")
+
+record(ao, "$(P)TIME:VOL:SP")
```

Comment:
Is there a reason the volume to pump is under `TIME:`? I would expect this PV name to follow that of `$(P)VOL:REMAINING`

Reply:
You're right. I refactored these under the prefix VOL. That way it's consistent for TIME and VOL as the 'bits' you're interrogating or changing.